dali/imgcodec/parsers/webp.cc

```diff
@@ -12,19 +13,158 @@
 // limitations under the License.
 
+#include <algorithm>
+#include <string>
 #include "dali/imgcodec/parsers/webp.h"
+#include "dali/core/byte_io.h"
 
 namespace dali {
 namespace imgcodec {
 
+namespace {
+
+// Chunk specification:
+// https://developers.google.com/speed/webp/docs/riff_container#riff_file_format
+//
+// Header layout:
+// RiffHeader | vp8_header_t | <WebpLossyHeader/WebpLosslessHeader>
+//
+// The WebpLossyHeader and WebpLosslessHeader are selected based on a identifier,
+// respectively "VP8 " and "VP8L" in vp8_header_t. "VP8X" in vp8_header_t signifies
+// WebP extended file format, which is not supported.
+
+// Struct must be packed for reading them to work correctly
+#pragma pack(push, 1)
+
+struct RiffHeader {
+  std::array<uint8_t, 4> riff_text;
+  uint32_t file_size;
+  std::array<uint8_t, 4> webp_text;
+};
+static_assert(sizeof(RiffHeader) == 12);
+
+using vp8_header_t = std::array<uint8_t, 4>;
+
+// Simple file format (lossy)
+// https://datatracker.ietf.org/doc/html/rfc6386#section-9.1
+struct WebpLossyHeader {
+  uint32_t chunk_size;
+  std::array<uint8_t, 3> frame_tag;
+  std::array<uint8_t, 3> sync_code;
+};
+static_assert(sizeof(WebpLossyHeader) == 10);
+
+// Simple file format (lossless)
+// https://developers.google.com/speed/webp/docs/webp_lossless_bitstream_specification#2_riff_header
+struct WebpLosslessHeader {
+  uint32_t chunk_size;
+  uint8_t signature_byte;
+};
+static_assert(sizeof(WebpLosslessHeader) == 5);
+
+#pragma pack(pop)  // end of packing scope
+
+template<size_t N>
+constexpr std::array<uint8_t, N - 1> tag(const char (&c)[N]) {
+  std::array<uint8_t, N - 1> a{};
+  std::copy(&c[0], &c[N - 1], a.begin());
+  return a;
+}
+
+bool is_valid_riff_header(const RiffHeader &header) {
+  return header.riff_text == tag("RIFF") && header.webp_text == tag("WEBP");
+}
+
+bool is_simple_lossy_format(const vp8_header_t &header) {
+  return header == tag("VP8 ");
+}
+
+bool is_simple_lossless_format(const vp8_header_t &header) {
+  return header == tag("VP8L");
+}
+
+bool is_extended_format(const vp8_header_t &header) {
+  return header == tag("VP8X");
+}
+
+bool is_sync_code_valid(const WebpLossyHeader &header) {
+  return header.sync_code == std::array<uint8_t, 3>{0x9D, 0x01, 0x2A};
+}
+
+bool is_sync_code_valid(const WebpLosslessHeader &header) {
+  return header.signature_byte == 0x2F;
+}
+
+template<size_t N>
+std::string sequence_of_integers(const std::array<uint8_t, N> &data) {
+  std::string result;
+  for (size_t i = 0; i < N; i++)
+    result += (i == 0 ? "" : " ") + std::to_string(data[i]);
+  return result;
+}
+
+}  // namespace
+
 ImageInfo WebpParser::GetInfo(ImageSource *encoded) const {
+  auto stream = encoded->Open();
   ImageInfo info;
-  // info.shape = {h, w, c};   // TODO(janton)
+
+  stream->Skip<RiffHeader>();
+  const auto vp8_header = stream->ReadOne<vp8_header_t>();
+  if (is_simple_lossy_format(vp8_header)) {
+    const auto lossy_header = stream->ReadOne<WebpLossyHeader>();
```

Comment:
```suggestion
    auto lossy_header = stream->ReadOne<WebpLossyHeader>();
    from_little_endian(lossy_header);
```

Reply:
I planned on adding this to the next [PR](https://github.com/staniewzki/DALI/tree/webp-exif) where I will actually use the chunk_size. Doing this now would involve a lot of rebasing and recompiling.

Reply:
ACK